java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -36,74 +36,95 @@ class ExpressionEvaluator(
             }
         }
     }
 
-    override fun visit(expr: Expression.WhereExpression): Any {
-        return (expr.qualifier.accept(this) as Sequence<*>).filter {
-            expr.expr.accept(this) == true
+    override fun visit(expr: Expression.WhereExpression, ctx: Scope): Any {
+        return (expr.qualifier.accept(this, ctx) as Sequence<Scope>).filter {
+            expr.expr.accept(this, it) == true
         }
     }
 
-    override fun visit(expr: Expression.LetExpression): Any {
-        return (expr.qualifier.accept(this) as Sequence<*>).map {
-            currentScope.set(expr.variableName, expr.variableExpr.accept(this))
+    override fun visit(expr: Expression.LetExpression, ctx: Scope): Any {
+        return (expr.qualifier.accept(this, ctx) as Sequence<Scope>).map { scope ->
+            scope.set(expr.variableName, expr.variableExpr.accept(this, scope))
         }
     }
 
-    override fun <T> visit(expr: Expression.SelectExpression<T>): Any {
-        return (expr.qualifier.accept(this) as Sequence<*>).map {
-            expr.expr.accept(this) as T
+    override fun <T> visit(expr: Expression.SelectExpression<T>, ctx: Scope): Any {
+        return (expr.qualifier.accept(this, ctx) as Sequence<Scope>).map {
+            expr.expr.accept(this, it) as T
         }
     }
 
-    override fun visit(expr: Expression.NewExpression): Any {
-        val newScope = scopeCreator(expr.schemaName.firstOrNull() ?: "")
-        expr.fields.forEach { (fieldName, fieldExpr) ->
-            newScope.set(fieldName, fieldExpr.accept(this))
-        }
-        return newScope
+    override fun visit(expr: Expression.NewExpression, ctx: Scope): Any {
+        val initScope = scopeCreator(expr.schemaName.firstOrNull() ?: "")
+        return expr.fields.fold(initScope.builder()) { builder, (fieldName, fieldExpr) ->
+           builder.set(fieldName, fieldExpr.accept(this, ctx))
+        }.build()
     }
 
-    override fun <T> visit(expr: Expression.FunctionExpression<T>): Any? {
-        val arguments = expr.arguments.map { it.accept(this) }.toList()
+    override fun <T> visit(expr: Expression.FunctionExpression<T>, ctx: Scope): Any? {
+        val arguments = expr.arguments.map { it.accept(this, ctx) }.toList()
         return expr.function.invoke(this, arguments)
     }
+
+    private fun compareBy(selector: Pair<Expression<Any>, Boolean>): Comparator<Scope> {
+        val comparator = { scope: Scope ->
+            selector.first.accept(this@ExpressionEvaluator, scope) as Comparable<Any>
+        }
+
+        return if (selector.second) {
+            compareByDescending(comparator)
+        } else {
+            compareBy(comparator)
+        }
+    }
+
+    override fun <T> visit(expr: Expression.OrderByExpression<T>, ctx: Scope): Any {
+        val nullComp: Comparator<Scope>? = null
+        return (expr.qualifier.accept(this, ctx) as Sequence<Scope>).sortedWith(
+            expr.selectors.fold(nullComp) { previous, selector ->
```

Comment:
I hope this part isn't getting too magical. Essentially, the evaluator works like this:

1. QualifiedExpressions produce a sequence of Scopes, which represent a snapshot or program memory via scope-chain
2. Visitors pass the scope to use to visit/accept methods
3. Whenever you need to 'eval' something, give a scope, usually from the previous qualifier sequence
4. OrderBy has to perform multiple passes over the Sequence<Scope>, each time it needs to compare two things, A and B, Kotlin invokes our comparator (bad name, should be selectorLambda or something) function which converts the Scope into a Comparable by evaluating the selector with respect to the Scope. That is, compare(scopeA, scopeB) -> compare(eval(expr, scopeA), eval(expr, scopeB))

5. the fold() and thenBy chaining just ensures the first lambda is converted to a comparator via compareBy(), and subsequent selectors (if more than 1) use .then(comparator2)

Reply:
I think it's quite understandable :) Dense, but not too magical :)